src/librustc_mir/build/matches/mod.rs

```diff
@@ -308,11 +309,7 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
                     block,
                     Statement {
                         source_info: ty_source_info,
-                        kind: StatementKind::AscribeUserType(
-                            place,
-                            ty::Variance::Invariant,
-                            user_ty,
-                        ),
+                        kind: StatementKind::AscribeUserType(place, variance, user_ty),
```

Comment:
I've updated the PR to have a `Option<ty::Variance>` so that - in every case other then associated constants - we can set this to `None` - I then use `.unwrap_or(..)` with `Invariant` or `Covariant` depending on what the code did previously. This means that we should be using the same variance as we did before, except for the associated constants case that this PR aims to fix where we override it.